nifi-nar-bundles/nifi-windows-event-log-bundle/nifi-windows-event-log-processors/src/main/java/org/apache/nifi/processors/windows/event/log/EvtSubscribe.java

```diff
@@ -0,0 +1,272 @@
+        this(null, null);
+    }
+
+    /**
+     * Constructor that allows injection of JNA interfaces
+     *
+     * @param wEvtApi event api interface
+     * @param kernel32 kernel interface
+     */
+    public EvtSubscribe(WEvtApi wEvtApi, Kernel32 kernel32) {
+        this.wEvtApi = wEvtApi == null ? loadWEvtApi() : wEvtApi;
+        this.kernel32 = kernel32 == null ? loadKernel32() : kernel32;
+        this.sessionFactoryReference = new AtomicReference<>();
+        this.renderedXMLs = new LinkedList<>();
+    }
+
+    /**
+     * Register subscriber via native call
+     *
+     * @param context the process context
+     */
+    @OnScheduled
+    public void subscribeToEvents(ProcessContext context) {
+        int maxEventQueueSize = context.getProperty(MAX_EVENT_QUEUE_SIZE).asInteger();
+        evtSubscribeCallback = new EventSubscribeXmlRenderingCallback(getLogger(), s -> {
+            ProcessSessionFactory processSessionFactory = sessionFactoryReference.get();
+            if (processSessionFactory == null) {
+                addRenderedXml(s, maxEventQueueSize);
+            } else {
+                ProcessSession session = processSessionFactory.createSession();
+                createAndTransferEventFlowFile(session, s);
+                session.commit();
+            }
+        }, context.getProperty(MAX_BUFFER_SIZE).asInteger(), wEvtApi, kernel32);
+        subscriptionHandle = wEvtApi.EvtSubscribe(null, null,
+                context.getProperty(CHANNEL).getValue(), context.getProperty(QUERY).getValue(), null, null,
+                evtSubscribeCallback, WEvtApi.EvtSubscribeFlags.SUBSCRIBE_TO_FUTURE);
+    }
+
+    /**
+     * Adds a rendered xml to the queue to be processed.
+     * Should only be called within block synchronized on evtSubscribeCallback before the sessionFactoryReference is set
+     *
+     * @param s the rendered xml string
+     * @param maxEventQueueSize the maximum size of the event queue
+     */
+    protected void addRenderedXml(String s, int maxEventQueueSize) {
+        renderedXMLs.add(s);
+        while (renderedXMLs.size() > maxEventQueueSize) {
+            renderedXMLs.poll();
+        }
+    }
+
+    /**
+     * Cleanup native subscription
+     */
+    @OnStopped
+    public void closeSubscriptionHandle() {
+        kernel32.CloseHandle(subscriptionHandle);
+        subscriptionHandle = null;
+    }
+
+    /**
+     * Creates a flow file from the given XML and transfers it to the success relationship
+     *
+     * @param session the process session
+     * @param xml the XML
+     */
+    protected void createAndTransferEventFlowFile(ProcessSession session, String xml) {
+        FlowFile flowFile = session.create();
+        flowFile = session.write(flowFile, out -> out.write(xml.getBytes(Charsets.UTF_8)));
+        flowFile = session.putAttribute(flowFile, CoreAttributes.MIME_TYPE.key(), MediaType.APPLICATION_XML_UTF_8.toString());
+        session.transfer(flowFile, REL_SUCCESS);
```

Comment:
In the event the session fails here, the session will roll back this flowfile and any already created flowfile (in this ontrigger). If this happens any rolled back flowfile will have already been removed from the queue and the data will be lost. You should first peek at the queue to get the first piece of data. Then create the flowfile and populate it. Then transfer it, commit  and remove it from the queue. This will ensure there is no data loss in the event of an error.

For an example, checkout the transferQueue method in ConsumeMQTT[1].

[1] https://github.com/apache/nifi/blob/master/nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java#L268